auslib/db.py

```diff
@@ -665,9 +718,243 @@ def rollbackChange(self, change_id, changed_by, transaction=None):
+        for row in self.select(where=where, transaction=transaction):
+            base_row = {col[5:]: row[col] for col in row if col.startswith("base_")}
+            base_table_where = {pk: row["base_%s" % pk] for pk in self.base_primary_key}
+            # TODO: What permissions *should* be required to delete a scheduled change?
+            # It seems a bit odd to be checking base table update/insert here. Maybe
+            # something broader should be required?
+            if base_row.get("base_data_version"):
+                self.baseTable.update(base_table_where, base_row, changed_by, base_row["base_data_version"], transaction=transaction, dryrun=True)
+            else:
+                self.baseTable.insert(changed_by, transaction=transaction, dryrun=True, **base_row)
+
+        if not dryrun:
+            return super(ScheduledChangeTable, self).delete(where, changed_by, old_data_version, transaction)
+
+    def enactChange(self, sc_id, enacted_by, transaction=None):
+        if not self.db.hasPermission(enacted_by, "scheduled_change", "enact", transaction=transaction):
+            raise PermissionDeniedError("%s is not allowed to enact scheduled changes", enacted_by)
+
+        sc = self.select(where=[self.sc_id == sc_id], transaction=transaction)[0]
+        what = {}
+        for col in sc:
+            if col.startswith("base_"):
+                what[col[5:]] = sc[col]
+
+        # The scheduled change is marked as complete first to avoid it being
+        # updated unnecessarily when the base table's update method calls
+        # mergeUpdate. If the base table update fails, this will get reverted
+        # when the transaction is rolled back.
+        self.update([self.sc_id == sc_id], {"complete": True}, changed_by=sc["scheduled_by"], old_data_version=sc["data_version"], transaction=transaction)
+
+        # If the scheduled change had a data version, it means the row already
+        # exists, and we need to use update() to enact it.
+        if what["data_version"]:
+            where = []
+            for col in self.base_primary_key:
+                where.append((getattr(self.baseTable, col) == sc["base_%s" % col]))
+            self.baseTable.update(where, what, sc["scheduled_by"], sc["base_data_version"], transaction=transaction)
+        else:
+            self.baseTable.insert(sc["scheduled_by"], transaction=transaction, **what)
+
+    def mergeUpdate(self, old_row, what, changed_by, transaction=None):
+        where = []
+        for col in self.base_primary_key:
+            where.append((getattr(self, "base_%s" % col) == old_row[col]))
+
+        scheduled_changes = self.select(where=where, transaction=transaction)
+        if not scheduled_changes:
+            self.log.debug("No scheduled changes found for update; nothing to do")
+            return
+        for sc in scheduled_changes:
+            if sc["complete"]:
```

Comment:
(moving this to the right place) In person we talked about moving the complete=True check into the select, so that the db does the work when the # of scheduled changes grows.